staging/src/k8s.io/client-go/rest/with_retry.go

```diff
@@ -147,34 +164,84 @@ func (r *withRetry) NextRetry(req *http.Request, resp *http.Response, err error,
 	//     need to check if it is retryable
 	seconds, wait := checkWait(resp)
 	if !wait {
-		return retryAfter, false
+		return false
 	}
 
-	retryAfter.Wait = time.Duration(seconds) * time.Second
-	retryAfter.Reason = getRetryReason(r.attempts, seconds, resp, err)
-	return retryAfter, true
+	r.retryAfter.Wait = time.Duration(seconds) * time.Second
+	r.retryAfter.Reason = getRetryReason(r.attempts, seconds, resp, err)
+	return true
 }
 
-func (r *withRetry) BeforeNextRetry(ctx context.Context, backoff BackoffManager, retryAfter *RetryAfter, url string, body io.Reader) error {
+func (r *withRetry) PrepareForNextRetry(ctx context.Context, request *Request) error {
 	// Ensure the response body is fully read and closed before
 	// we reconnect, so that we reuse the same TCP connection.
+	//
+	// r.retryAfter represents the retry after parameters
+	// for the next attempt to be made
+	if r.retryAfter == nil {
+		return fmt.Errorf("no retry-after parameters set")
+	}
 	if ctx.Err() != nil {
 		return ctx.Err()
 	}
 
-	if seeker, ok := body.(io.Seeker); ok && body != nil {
+	if seeker, ok := request.body.(io.Seeker); ok && request.body != nil {
 		if _, err := seeker.Seek(0, 0); err != nil {
-			return fmt.Errorf("can't Seek() back to beginning of body for %T", r)
+			return fmt.Errorf("can't Seek() back to beginning of body for %T", request)
+		}
+	}
+
+	klog.V(4).Infof("Got a Retry-After %s response for attempt %d to %v", r.retryAfter.Wait, r.retryAfter.Attempt, request.URL().String())
+	return nil
+}
+
+func (r *withRetry) Before(ctx context.Context, request *Request) error {
+	// r.retryAfter represents the retry after parameters for the
+	// next attempt to be made immediately after this function returns.
+	url := request.URL()
+	if request.backoff != nil {
+		if r.retryAfter != nil {
+			// TODO(tkashem) with default set to use exponential backoff
+			//  we can merge these two sleeps:
+			//  BackOffManager.Sleep(max(backoffManager.CalculateBackoff(), retryAfter))
+			//  see https://github.com/kubernetes/kubernetes/issues/108302
+			request.backoff.Sleep(r.retryAfter.Wait)
 		}
+
+		// we always do a backoff sleep including the first try
+		request.backoff.Sleep(request.backoff.CalculateBackoff(url))
```

Comment:
does it simplify something if we sum the values ?
request.backoff.CalculateBackoff(url) + r.retryAfter.Wait

Reply:
I think the suggestion from @wojtek-t was to `max(backoffManager.CalculateBackoff(), retryAfter)`, this will be done in a follow up PR, this PR is refactor-only